### PR TITLE
snmp destination driver: make it threaded

### PR DESCRIPTION
**Pull request by norberttak**
Refactor snmp destination driver to be based on threaded destination driver.
During the development it turned out that we use g_atomic_int_exchange_and_add in seqnum.h which is deprecated since glib version 2.30. A commit has been created to fix this deprecated function.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/snmp-dest/snmpdest.c [hunk 1]
```diff
@@ -819,36 +694,55 @@ snmpdest_dd_free(LogPipe *d)
...
+  /* prepare the host:port format - the session.remote_port doesn't work */
+  self->host_port = g_string_sized_new(64);
 
-  g_mutex_unlock(self->queue_mutex);
+  if (self->transport)
+    g_string_append_printf(self->host_port, "%s:", self->transport);
+  g_string_append_printf(self->host_port, "%s:%d", self->host, self->port);
 
-  g_cond_signal(self->writer_thread_wakeup_cond);
+  msg_verbose("Initializing SNMP destination worker thread",
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: Removed this diag message

File: modules/snmp-dest/snmpdest.c [hunk 2]
```diff
@@ -819,36 +694,55 @@ snmpdest_dd_free(LogPipe *d)
...
+              evt_tag_str("host", self->host),
+              evt_tag_int("port", self->port));
+
+  gchar *tz = self->template_options.time_zone[LTZ_LOCAL];
+  gchar *time_zone = cfg->template_options.time_zone[LTZ_SEND];
+  if (!tz && time_zone)
+    /* inherit the global set_time_zone if the local one hasn't been set */
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: comment moved above the "if"

File: modules/snmp-dest/snmpdest.c [hunk 3]
```diff
@@ -437,7 +427,7 @@ snmpdest_worker_insert(SNMPDestDriver *self, LogMessage *msg, LogPathOptions *pa
           msg_warning("SNMP: error adding variable",
                       evt_tag_str("value", fs->str));
           log_msg_unref(msg);
-          return FALSE;
+          return LTR_ERROR;
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: added teh missing free functions: 
g_string_free(fs, TRUE);
snmp_free_pdu(pdu);

File: modules/snmp-dest/snmpdest.h [hunk 4]
```diff
@@ -59,15 +60,7 @@ typedef struct
   gchar *enc_password;
   gchar *transport;
 
-  StatsCounterItem *dropped_messages;
-  StatsCounterItem *queued_messages;
-  StatsCounterItem *processed_messages;
-
   GMutex *queue_mutex;
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.h]: removed

File: modules/snmp-dest/snmpdest.c [hunk 5]
```diff
@@ -819,36 +694,55 @@ snmpdest_dd_free(LogPipe *d)
...
-  g_mutex_lock(self->queue_mutex);
+  GlobalConfig *cfg = log_pipe_get_config((LogPipe *)d);
 
-  stats_counter_inc(self->processed_messages);
-  log_queue_push_tail(self->queue, msg, path_options);
+  /* prepare the host:port format - the session.remote_port doesn't work */
+  self->host_port = g_string_sized_new(64);
 
-  g_mutex_unlock(self->queue_mutex);
+  if (self->transport)
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: I don't have start/reload functions so I keep this code in the thread_init function but added a check if self->host_port exists then don't create it again.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
Can we merge the two snmp related plugins?

**Comment by norberttak** [comment]
Is it a good idea to merge the snmp trap (which is a parser for a file source) and the snmp destination into a common driver?

**Comment by bazsi** [comment]
I think it is. The other being a parser for snmptrapd format is an
implementation detail, that might change eventually. The common point is
snmp, and we tend to include the source/destination for a specific
transport into the same module. Their collocation into the same module
might open up code sharing in the future.

Think of MIB management for instance

File: modules/snmp-dest/snmpdest.c [hunk 6]
```diff
@@ -705,78 +623,38 @@ static gboolean
...
-  if (self->transport)
-    g_string_append_printf(self->host_port, "%s:", self->transport);
-  g_string_append_printf(self->host_port, "%s:%d", self->host, self->port);
-
-  self->queue = log_dest_driver_acquire_queue(&self->super, snmpdest_dd_format_persist_name(self));
-
-  guint SCS_SNMP = stats_register_type("snmp");
-
-  stats_lock();
-  stats_cluster_logpipe_key_set(&self->sc_key_queued, SCS_SNMP | SCS_DESTINATION,
```
Comment (szemere) [inline modules/snmp-dest/snmpdest.c]: I think it is worth to mention, that statistics output will change with this change.

1. `LogThreadedDestDriver` registers the following statistics:
```
SC_TYPE_DROPPED
SC_TYPE_PROCESSED
SC_TYPE_WRITTEN
```
(and we got `SC_TYPE_QUEUED` via logqueue)

2. I tried to test the original statistics on the latest master. They did not work for me.
 Thus I would not consider this change as an "error" (quite the opposite), but I will mark it as "user visible"

Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: I created a ticket for the techwriters to document the different behavior.

File: modules/snmp-dest/snmpdest.c [hunk 7]
```diff
@@ -393,26 +394,15 @@ sanitize_fs(GString *fs, gint code)
...
-snmpdest_worker_insert(SNMPDestDriver *self, LogMessage *msg, LogPathOptions *path_options)
+static LogThreadedResult
+snmpdest_worker_insert(LogThreadedDestDriver *s, LogMessage *msg)
 {
+  SNMPDestDriver *self = (SNMPDestDriver *) s;
   netsnmp_pdu *pdu;
 
-  if (!self->session_initialized)
-    {
-      if (!snmpdest_dd_session_init(self))
```
[the remark this answers is not in the document]
Comment (norberttak) [reply modules/snmp-dest/snmpdest.c]: obsolete forward declaration removed

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**